src/Compilers/CSharp/Portable/Compiler/MethodCompiler.cs

```diff
@@ -930,7 +930,7 @@ public override object VisitField(FieldSymbol symbol, TypeCompilationState argum
                     var lazySemanticModel = body == null ? null : new Lazy<SemanticModel>(() =>
                     {
                         var syntax = body.Syntax;
-                        var semanticModel = (CSharpSemanticModel)_compilation.GetSemanticModel(syntax.SyntaxTree);
+                        var semanticModel = (CSharpSemanticModel)AnalyzerDriver.GetOrCreateCachedSemanticModel(syntax.SyntaxTree, _compilation, _cancellationToken);
```

Comment:
@gafter The issue is that the semantic model stored onto a CompilationEvent is weakly referenced. For IDE case, a compilation event can be alive for lot longer then command line case - you might execute some analyzers on a symbol and return back to completing analyzing the event later. So we basically use IDE's semantic model to force bind symbols in a document, and pass that onto the analyzer. I verified the perf numbers from this PR that allocations reduce quite a bit during typing, but seem to shoot up during command line build. The later is likely due to the issue you pointed out at the top of this thread.
I am going to try to create a PR for the change I suggested above and get perf numbers for it and see if it helps or not.